src/test/java/com/uber/cadence/internal/common/RetryerTest.java

```diff
@@ -105,4 +106,29 @@ public void testInterruptedException() throws InterruptedException {
     }
     assertTrue(System.currentTimeMillis() - start < 100000);
   }
+
+  @Test
+  public void testMaxAttempt() throws InterruptedException {
+    RetryOptions options =
+        new RetryOptions.Builder()
+            .setInitialInterval(Duration.ofMillis(10))
+            .setMaximumInterval(Duration.ofMillis(100))
+            .setExpiration(Duration.ofMillis(500))
+            .setMaximumAttempts(1)
+            .validateBuildWithDefaults();
+    long start = System.currentTimeMillis();
+    try {
+      Retryer.retryWithResultAsync(
+              options,
+              () -> {
+                throw new IllegalArgumentException("simulated");
+              })
+          .get();
+      fail("unreachable");
+    } catch (ExecutionException e) {
+      assertTrue(e.getCause() instanceof IllegalArgumentException);
+      assertEquals("simulated", e.getCause().getMessage());
+    }
+    assertTrue(System.currentTimeMillis() - start < 500);
```

Comment:
Instead of assert on time, increment a counter and assert on attempt?

Reply:
that is ideal but may need to update `retryWithResultAsync` signature to return attempts, since there is not way to have a counter in java closure. I will leave it for now.